arangod/IResearch/GeoAnalyzer.cpp

```diff
@@ -21,266 +21,338 @@
-                                     values::default_constructed_value<
-                                         std::initializer_list<std::string>>>,
-        factory_deserialized_default<LONGITUDE_PARAM, StringVectorDeserializer,
-                                     values::default_constructed_value<
-                                         std::initializer_list<std::string>>>>>;
-
-using ValidatingGeoPointsOptionsDeserializer =
-    validate<GeoPointsOptionsDeserializer, GeoPointAnalyzerOptionsValidator>;
+  r = get(kLongitudeParam, options.longitude);
+  if (!r.ok()) {
+    return r;
+  }
+  if (options.latitude.empty() != options.longitude.empty()) {
+    options.latitude = {};
+    options.longitude = {};
+    return {TRI_ERROR_BAD_PARAMETER,
+            absl::StrCat("'", kLatitudeParam, "' and '", kLongitudeParam,
+                         "' should be both empty or non-empty.")};
+  }
+  return {};
+}
 
-template<typename Analyzer>
-struct Deserializer;
-
-template<>
-struct Deserializer<GeoJSONAnalyzer> {
-  using type = GeoJSONOptionsDeserializer;
-};
-
-template<>
-struct Deserializer<GeoPointAnalyzer> {
-  using type = ValidatingGeoPointsOptionsDeserializer;
-};
-
-void toVelocyPack(VPackBuilder& builder, GeoOptions const& opts) {
-  VPackObjectBuilder root(&builder);
-  builder.add(MAX_CELLS_PARAM, VPackValue(opts.maxCells));
-  builder.add(MIN_LEVEL_PARAM, VPackValue(opts.minLevel));
-  builder.add(MAX_LEVEL_PARAM, VPackValue(opts.maxLevel));
-}
-
-void toVelocyPack(VPackBuilder& builder, GeoJSONAnalyzer::Options const& opts) {
-  VPackObjectBuilder root(&builder);
-  switch (opts.type) {
-    case GeoJSONAnalyzer::Type::SHAPE:
-      builder.add(TYPE_PARAM, VPackValue(GEOJSON_SHAPE_TYPE));
-      break;
-    case GeoJSONAnalyzer::Type::CENTROID:
-      builder.add(TYPE_PARAM, VPackValue(GEOJSON_CENTROID_TYPE));
-      break;
-    case GeoJSONAnalyzer::Type::POINT:
-      builder.add(TYPE_PARAM, VPackValue(GEOJSON_POINT_TYPE));
-      break;
+Result fromVelocyPack(velocypack::Slice object,
+                      GeoVPackAnalyzer::Options& options) {
+  auto r = fromVelocyPackBase(object, options);
+  if (!r.ok()) {
+    return r;
   }
-  builder.add(VPackValue(OPTIONS_PARAM));
-  toVelocyPack(builder, opts.options);
+  return getBool(object, kLegacyParam, options.legacy);
 }
 
-void toVelocyPack(VPackBuilder& builder,
-                  GeoPointAnalyzer::Options const& opts) {
-  auto addArray = [&builder](const char* name,
-                             std::vector<std::string> const& values) {
-    VPackArrayBuilder arrayScope(&builder, name);
-    for (auto& value : values) {
-      builder.add(VPackValue(value));
+Result fromVelocyPack(velocypack::Slice object,
+                      GeoS2Analyzer::Options& options) {
+  auto r = fromVelocyPackBase(object, options);
+  if (r.ok()) {
+    return r;
+  }
+  auto value = object.get(kCompressionParam);
```

Comment:
Suggest to have meaningful names like "fast"/"compact" here.  As 0/1 is not very clear for the end-user.

Reply:
I will think about it

Reply:
In general I just don't really like such names

Reply:
I think in next PR